### DISCUSSION TITLE
Enhance typescript type check

### DISCUSSION THREAD

**Discussion by vanenshi** [mobxjs/mobx#2867] [Q&A] [answered]
as you know, using the `@inject` with typescript always cause some problem for the programmers (see mobxjs/mobx#1778 and mobxjs/mobx-react#256)

so I decided to use a custom decorator for both `inject`  and class component `props` type checking
I use the  solution mentioned [here](https://github.com/mobxjs/mobx-react/issues/256#issuecomment-467196759)
```ts
import { Component } from 'react';
import { inject } from 'mobx-react';
import { Stores } from '@Stores'; // root store inteface

export class ComponentWithStore<IStores, IProps = {}, IStates = {}> extends Component<IProps, IStates> {
  public get stores() {
    return (this.props as any) as IStores;
  }
}

export const connect = (...args: Array<keyof Stores>) => inject(...args);
```
how about enhancing the `@inject` decorator with this??

**Comment by mweststrate** [comment, marked as answer]
inject is deprecated and we recommend to use React.createContext instead,
which is strongly typed ootb.

**Reply by vanenshi** [reply]
Thank you

**Comment by iChenLei** [comment]
I think it's time to finish https://github.com/mobxjs/mobx/issues/2541 work, we can add dev warn firstly for progressive migrate and then separate `Provider`/`inject` from package .
